File: services/audio_service.py
from models.audio import Audio
from models.song import Song
from services.player_service import PlayerService


class AudioService:

    def __init__(self, song: Song, callback):
        self._song = song
        self._index = 0
        self._callback = callback
        self._players = {}
        self._player = None
        self._preload_players()
        self._select_current_audio()

    def forward(self):
        
        self._index = self._index+1

        self._index = 0 if self._index == len(
            self._song.audios) else self._index

        self._select_current_audio()

    def current(self) -> Audio :
        return self._song.audios[self._index]

    def play(self):
        self._player.play()

    def stop(self):
        self._player.stop()

    def autoforward(self):
        if self._song.autoforward:
            self.forward()

    def _select_current_audio(self):
        current = self.current()

        if self._player is not None:
            self._player.stop()

        self._player = self._players[current.id]

    def _preload_players(self):
        for audio in self._song.audios:
            self._players[audio.id] = PlayerService(
                audio,
                self._callback,
                self.autoforward
            )


Audio_Service = AudioService

File: src/audio_service.cpp
#include "audio_service.hpp"

namespace cherry {

AudioService::AudioService(
    Song song,
    AudioPlayerFactory& player_factory,
    PlaybackSequencer& sequencer,
    PlayerService::EventCallback callback,
    LoadProgressCallback load_progress)
    : song_(std::move(song)) {
    const std::size_t total = song_.audios.size();
    players_.reserve(total);

    for (const auto& audio : song_.audios) {
        players_.push_back(std::make_unique<PlayerService>(
            audio,
            player_factory.create(),
            sequencer,
            callback,
            [this]() { autoforward(); }));

        if (load_progress) {
            load_progress(players_.size(), total);
        }
    }

    std::lock_guard<std::mutex> lock(mutex_);
    select_current_audio_locked();
}

void AudioService::select_current_audio_locked() {
    if (current_player_ != nullptr) {
        current_player_->stop();
    }

    current_player_ = players_[index_].get();
}

void AudioService::advance_locked() {
    ++index_;
    if (index_ == song_.audios.size()) {
        index_ = 0;
    }

    select_current_audio_locked();
}

void AudioService::forward() {
    std::lock_guard<std::mutex> lock(mutex_);
    advance_locked();
}

void AudioService::autoforward() {
    std::lock_guard<std::mutex> lock(mutex_);
    if (song_.autoforward) {
        advance_locked();
    }
}

const Audio& AudioService::current() const {
    std::lock_guard<std::mutex> lock(mutex_);
    return song_.audios[index_];
}

PlayerService* AudioService::current_player() const {
    std::lock_guard<std::mutex> lock(mutex_);
    return current_player_;
}

void AudioService::play() {
    current_player()->play();
}

void AudioService::stop() {
    current_player()->stop();
}

double AudioService::progress() const {
    return current_player()->progress();
}

}  // namespace cherry
